src/AzureClient/AzureClient.cs

```diff
@@ -346,11 +341,41 @@ public async Task<ExecutionResult> GetConnectionStatusAsync(IChannel? channel, C
                 return AzureClientError.InvalidEntryPoint.ToExecutionResult();
             }
 
+            channel?.Stdout($"Submitting {submissionContext.OperationName} to target {ActiveTarget.TargetId}...");
+
             try
             {
+                // QirSubmitter and CreateMachine have return types with different base types
+                // but both have a SubmitAsync method that returns an IQuantumMachineJob.
+                // Thus, we can branch on whether we need a QIR submitter or a translator,
+                // but can use the same task object to represent both return values.
+                Task<Runtime.IQuantumMachineJob>? jobTask = null;
+                if (this.ActiveTarget.TargetId.StartsWith(MicrosoftSimulator))
+                {
+                    var submitter = SubmitterFactory.QirSubmitter(this.ActiveTarget.TargetId, this.ActiveWorkspace, this.StorageConnectionString);
+                    if (submitter == null)
+                    {
+                        // We should never get here, since ActiveTarget should have already been validated at the time it was set.
+                        channel?.Stderr($"Unexpected error while preparing job for execution on target {ActiveTarget.TargetId}.");
+                        return AzureClientError.InvalidTarget.ToExecutionResult();
+                    }
+                    jobTask = entryPoint.SubmitAsync(submitter, submissionContext);
+                }
+                else
+                {
+                    var machine = AzureFactory.CreateMachine(this.ActiveWorkspace, this.ActiveTarget.TargetId, this.StorageConnectionString);
+                    if (machine == null)
+                    {
+                        // We should never get here, since ActiveTarget should have already been validated at the time it was set.
+                        channel?.Stderr($"Unexpected error while preparing job for execution on target {ActiveTarget.TargetId}.");
+                        return AzureClientError.InvalidTarget.ToExecutionResult();
+                    }
+                    jobTask = entryPoint.SubmitAsync(machine, submissionContext);
+                }
```

Comment:
Since branching here depends on whether a `IQirSubmitter` or a `IQuantumMachine` object is available for the particular target, I believe it would be better to organize this part as:
```csharp
if (SubmitterFactory.QirSubmitter(this.ActiveTarget.TargetId, this.ActiveWorkspace, this.StorageConnectionString) is IQirSubmitter submitter)
{
    // Continue with submitter logic.
}
else if (AzureFactory.CreateMachine(this.ActiveWorkspace, this.ActiveTarget.TargetId, this.StorageConnectionString) is IQuantumMachine machine)
{
    // Continue with machine logic.
}
```

Reply:
I'm not quite following your suggested change. Why would we want to call into the SubmitterFactory if we don't want to deal with QIR submitters? Shouldn't the deciding logic be external to those calls (SubmitterFactory.QirSubmitter and AzureFactory.CreateMachine)? If we are calling a factory function we should already be expecting that it will return a valid object, right?

Reply:
In the way the logic is currently written, we decide whether we need an `IQirSubmitter` or an `IQuantumMachine` object based on the target ID. However, both the `SubmitterFactory.QirSubmitter` and the `AzureFactory.CreateMachine` factory methods already have logic inside them to handle this. If the target ID is not valid to create a submitter object or a machine object, they return null.

My suggestion is to leverage this logic. If a submitter object for the target can be created, then we can just use it. Otherwise, we try to create a machine object for the target and use it if it could be created.

Hopefully this clarifies my comment.